File: api/src/test/java/net/blueshell/api/validation/GuestOrUserRequiredValidatorTest.java
package net.blueshell.api.validation;

import jakarta.validation.ConstraintValidatorContext;
import net.blueshell.api.dto.event.EventSignUpDTO;
import net.blueshell.api.factory.dto.GuestDTOFactory;
import net.blueshell.api.factory.dto.event.EventSignUpDTOFactory;
import net.blueshell.api.factory.dto.user.SimpleUserDTOFactory;
import net.blueshell.api.validation.event.GuestOrUserRequiredValidator;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.*;

/**
 * Unit tests for GuestOrUserRequiredValidator.
 */
@SpringBootTest
class GuestOrUserRequiredValidatorTest {

    private final GuestOrUserRequiredValidator validator = new GuestOrUserRequiredValidator();

    private final EventSignUpDTOFactory signUpFactory;
    private final GuestDTOFactory guestFactory;
    private final SimpleUserDTOFactory simpleUserFactory;

    @Autowired
    GuestOrUserRequiredValidatorTest(
            EventSignUpDTOFactory signUpFactory,
            GuestDTOFactory guestFactory,
            SimpleUserDTOFactory simpleUserFactory
    ) {
        this.signUpFactory = signUpFactory;
        this.guestFactory = guestFactory;
        this.simpleUserFactory = simpleUserFactory;
    }

    @Test
    void valid_when_guest_present() {
        EventSignUpDTO dto = signUpFactory.createWithCustomizations(es -> {
            es.setEventId(1L);
            es.setGuest(guestFactory.createWithCustomizations(g -> g.setName("Guesty McGuestface")));
            es.setUser(null);
            es.setUserId(null);
        });

        assertTrue(validator.isValid(dto, mock(ConstraintValidatorContext.class)));
    }

    @Test
    void valid_when_user_present() {
        EventSignUpDTO dto = signUpFactory.createWithCustomizations(es -> {
            es.setEventId(1L);
            es.setUser(simpleUserFactory.createWithCustomizations(u -> u.setId(42L)));
            es.setUserId(null);
            es.setGuest(null);
        });

        assertTrue(validator.isValid(dto, mock(ConstraintValidatorContext.class)));
    }

    @Test
    void valid_when_userId_present() {
        EventSignUpDTO dto = signUpFactory.createWithCustomizations(es -> {
            es.setEventId(1L);
            es.setUser(null);
            es.setGuest(null);
            es.setUserId(99L);
        });

        assertTrue(validator.isValid(dto, mock(ConstraintValidatorContext.class)));
    }

    @Test
    void invalid_when_neither_guest_nor_user_provided() {
        EventSignUpDTO dto = signUpFactory.createWithCustomizations(es -> {
            es.setEventId(1L);
            es.setGuest(null);
            es.setUser(null);
            es.setUserId(null);
        });

        var ctx = mock(ConstraintValidatorContext.class, RETURNS_DEEP_STUBS);

        assertFalse(validator.isValid(dto, ctx));
        verify(ctx).disableDefaultConstraintViolation();
        verify(ctx).buildConstraintViolationWithTemplate("Either guest or user (or userId) must be provided.");
    }
}

File: api/src/test/kotlin/net/blueshell/api/event/api/validation/GuestOrUserRequiredValidatorTest.kt
package net.blueshell.api.event.api.validation

import jakarta.validation.ConstraintValidatorContext
import net.blueshell.api.event.api.dto.EventSignUpDTO
import net.blueshell.api.factory.dto.GuestDTOFactory
import net.blueshell.api.factory.dto.event.EventSignUpDTOFactory
import net.blueshell.api.factory.dto.user.SimpleUserDTOFactory
import net.blueshell.api.event.api.validation.GuestOrUserRequiredValidator
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.kotlin.mock
import org.mockito.kotlin.verify
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest

/**
 * Unit tests for GuestOrUserRequiredValidator.
 */
@SpringBootTest
class GuestOrUserRequiredValidatorTest @Autowired constructor(
    private val signUpFactory: EventSignUpDTOFactory,
    private val guestFactory: GuestDTOFactory,
    private val simpleUserFactory: SimpleUserDTOFactory
) {

    private val validator = GuestOrUserRequiredValidator()

    @Test
    fun `valid when guest present`() {
        val dto: EventSignUpDTO = signUpFactory.createWithCustomizations {
            it.eventId = 1L
            it.guest = guestFactory.createWithCustomizations { guest -> guest.name = "Guesty McGuestface" }
            it.user = null
            it.userId = null
        }

        assertTrue(validator.isValid(dto, mock<ConstraintValidatorContext>()))
    }

    @Test
    fun `valid when user present`() {
        val dto: EventSignUpDTO = signUpFactory.createWithCustomizations {
            it.eventId = 1L
            it.user = simpleUserFactory.createWithCustomizations { user -> user.id = 42L }
            it.userId = null
            it.guest = null
        }

        assertTrue(validator.isValid(dto, mock<ConstraintValidatorContext>()))
    }

    @Test
    fun `valid when userId present`() {
        val dto: EventSignUpDTO = signUpFactory.createWithCustomizations {
            it.eventId = 1L
            it.user = null
            it.guest = null
            it.userId = 99L
        }

        assertTrue(validator.isValid(dto, mock<ConstraintValidatorContext>()))
    }

    @Test
    fun `invalid when neither guest nor user provided`() {
        val dto: EventSignUpDTO = signUpFactory.createWithCustomizations {
            it.eventId = 1L
            it.guest = null
            it.user = null
            it.userId = null
        }

        val ctx = mock<ConstraintValidatorContext>(defaultAnswer = Mockito.RETURNS_DEEP_STUBS)

        assertFalse(validator.isValid(dto, ctx))
        verify(ctx).disableDefaultConstraintViolation()
        verify(ctx).buildConstraintViolationWithTemplate("Either guest or user (or userId) must be provided.")
    }
}
